mips-linux-gnu: relocation truncated to fit: R_MIPS_TLS_LDM

 http://sourceware.org/bugzilla/show_bug.cgi?id=12637

===================================================================
RCS file: /cvs/src/src/bfd/ChangeLog,v
retrieving revision 1.5298
retrieving revision 1.5299
diff -u -r1.5298 -r1.5299
--- src/bfd/ChangeLog	2011/04/09 16:48:36	1.5298
+++ src/bfd/ChangeLog	2011/04/10 11:10:30	1.5299
@@ -1,3 +1,9 @@
+2011-04-10  Richard Sandiford  <rdsandiford@googlemail.com>
+
+	PR ld/12637
+	* elfxx-mips.c (mips_elf_merge_got_with): Use arg->global_count
+	as the number of global entries when merging with the primary GOT.
+
 2010-11-04  Maciej W. Rozycki  <macro@codesourcery.com>
 
 	PR ld/10144
===================================================================
RCS file: /cvs/src/src/bfd/elfxx-mips.c,v
retrieving revision 1.282
retrieving revision 1.283
diff -u -r1.282 -r1.283
--- src/bfd/elfxx-mips.c	2011/01/22 10:16:28	1.282
+++ src/bfd/elfxx-mips.c	2011/04/10 11:10:29	1.283
@@ -4080,14 +4080,18 @@
   if (estimate >= from->page_gotno + to->page_gotno)
     estimate = from->page_gotno + to->page_gotno;
 
-  /* And conservatively estimate how many local, global and TLS entries
+  /* And conservatively estimate how many local and TLS entries
      would be needed.  */
-  estimate += (from->local_gotno
-	       + from->global_gotno
-	       + from->tls_gotno
-	       + to->local_gotno
-	       + to->global_gotno
-	       + to->tls_gotno);
+  estimate += from->local_gotno + to->local_gotno;
+  estimate += from->tls_gotno + to->tls_gotno;
+
+  /* If we're merging with the primary got, we will always have
+     the full set of global entries.  Otherwise estimate those
+     conservatively as well.  */
+  if (to == arg->primary)
+    estimate += arg->global_count;
+  else
+    estimate += from->global_gotno + to->global_gotno;
 
   /* Bail out if the combined GOT might be too big.  */
   if (estimate > arg->max_count)
-- 
